README.md
### Dictdiffer

Dictdiffer is a helper module that helps you to diff and patch dictionaries.

#### dictdiffer.diff
Compares two dictionary object and returns a diff result.

#### dictdiffer.patch
Applies the diff result to the old dictionary.

#### dictdiffer.swap
Swaps the diff result to revert changes.

#### dictdiffer.revert
A shortcut function that swaps and patches the diff result to
destination dictionary.

### Installation

    $ pip install dictdiffer

### Examples

```python
from dictdiffer import diff, patch, swap, revert

first = {
    "title": "hello",
    "fork_count": 20,
    "stargazers": ["/users/20", "/users/30"],
    "settings": {
        "assignees": [100, 101, 201],
    }
}


second = {
    "title": "hellooo",
    "fork_count": 20,
    "stargazers": ["/users/20", "/users/30", "/users/40"],
    "settings": {
        "assignees": [100, 101, 202],
    }
}

result = diff(first, second)

assert list(result) == [
    ('push', 'settings.assignees', [202]),
    ('pull', 'settings.assignees', [201]),
    ('push', 'stargazers', ['/users/40']),
    ('change', 'title', ('hello', 'hellooo'))]
```

Now we can apply the diff result with `patch` method.

```python
result = diff(first, second)
patched = patch(result, first)

assert patched == second
```

Also we can swap the diff result with `swap` method.

```python
result = diff(first, second)
swapped = swap(result)

assert list(swapped) == [
    ('pull', 'settings.assignees', [202]),
    ('push', 'settings.assignees', [201]),
    ('pull', 'stargazers', ['/users/40']),
    ('change', 'title', ('hellooo', 'hello'))]
```

Let's revert the last changes.

```python
reverted = revert(result, patched)
assert reverted == first
```

### Why I built it?

I decided to make revision history mechanism in [dbpatterns.com](http://dbpatterns.com) for documents. Documents are storing in MongoDB, 
so I was needed a dictionary differ that diffs two dictionary and reverts back changes.

### Contribution & Tests

After your changes, you could run the tests to ensure everything is
operating correctly.

    $ source run_tests.sh
    Ran 15 tests in 0.002s

    Name         Stmts   Miss  Cover   Missing
    ------------------------------------------
    dictdiffer      74      0   100%
    ------------------------------------------
    TOTAL           74      0   100%


CHANGES
Changes
=======

Version 0.9.0 (released 2021-07-22)

- Adds absolute tolerance feature for floats  (@adrien-berchet) (#152)
- Drops support of Python<3.5  (@adrien-berchet) (#160)
- Adds `assert_no_diff` helper to assist pytest users  (@joesolly) (#153)
- Migrates CI to gh-actions (@ParthS007 @diegodelemos) (#145)
- Removes dependency on pkg_resources (@eldruin)

Version 0.8.1 (released 2019-12-13)

- Fix invalid diff output for sets. (@jirikuncar @danielduhh)  (#133 #134)

Version 0.8.0 (released 2019-03-17)

- Respect `dot_notation` flag in ignore argument (@yoyonel) (#107)
- Adds argument for toggling dot notation in diff. (@robinchew)

Version 0.7.2 (released 2019-02-22)

- Two NaN values are considered the same, hence they are not shown in `diff`
  output.  (#114) (@t-b)
- Refactors `diff` method to reduce recursive call stack size.  (#112)
  (@yoyonel)
- Python porting best practice use feature detection instead
  of version detection to save an import and pass both PyLint
  and Flake8 tests with neither 'pragma' nor 'noqa'.  (@cclauss)

Version 0.7.1 (released 2018-05-04)

- Resolves issue with keys containing dots.  (#101)

Version 0.7.0 (released 2017-10-16)

- Fixes problem with diff results that reference the original structure by
  introduction of `deepcopy` for all possibly unhashable items. Thus the diff
  does not change later when the diffed structures change.
- Adds new option for patching and reverting patches in-place.
- Adds Python 3.6 to test matrix.
- Fixes the `ignore` argument when it contains a unicode value.

Version 0.6.1 (released 2016-11-22)

- Changes order of items for REMOVE section of generated patches when
  `swap` is called so the list items are removed from the end. (#85)
- Improves API documentation for `ignore` argument in `diff` function.
  (#79)
- Executes doctests during PyTest invocation.

Version 0.6.0 (released 2016-06-22)

- Adds support for comparing NumPy arrays.  (#68)
- Adds support for comparing mutable mappings, sequences and sets from
  `collections.abs` module.  (#67)
- Updates package structure, sorts imports and runs doctests.
- Fixes order in which handled conflicts are unified so that the
  Merger's unified_patches can be always applied.

Version 0.5.0 (released 2016-01-04)

- Adds tolerance parameter used when user wants to treat closed values
  as equals
- Adds support for comparing numerical values and NaN. (#54) (#55)

Version 0.4.0 (released 2015-03-11)

- Adds support for diffing and patching of sets. (#44)
- New tests for diff on the same lists. (#48)
- Fix for exception when dict has unicode keys and ignore parameter is
  provided. (#50)
- PEP8 improvements.

Version 0.3.0 (released 2014-11-05)

- Adds ignore argument to `diff` function that allows skipping check
  on specified keys. (#34 #35)
- Fix for diffing of dict or list subclasses. (#37)
- Better instance checking of diffing objects. (#39)

Version 0.2.0 (released 2014-09-29)

- Fix for empty list instructions. (#30)
- Regression test for empty list instructions.

Version 0.1.0 (released 2014-09-01)

- Fix for list removal issues during patching caused by wrong
  iteration. (#10)
- Fix for issues with multiple value types for the same key. (#10)
- Fix for issues with strings handled as iterables. (#6)
- Fix for integer keys. (#12)
- Regression test for complex dictionaries. (#4)
- Better testing with Github actions, tox, pytest, code coverage. (#10)
- Initial release of documentation on ReadTheDocs. (#21 #24)
- Support for Python 3. (#15)

Version 0.0.4 (released 2014-01-04)

- List diff behavior treats lists as lists instead of sets. (#3)
- Differed typed objects are flagged as `changed` now.
- Swap function refactored.

Version 0.0.3 (released 2013-05-26)

- Initial public release on PyPI.


dictdiffer/__init__.py
# This file is part of Dictdiffer.
#
# Copyright (C) 2013 Fatih Erikli.
# Copyright (C) 2013, 2014, 2015, 2016 CERN.
# Copyright (C) 2017-2019 ETH Zurich, Swiss Data Science Center, Jiri Kuncar.
#
# Dictdiffer is free software; you can redistribute it and/or modify
# it under the terms of the MIT License; see LICENSE file for more
# details.

"""Dictdiffer is a helper module to diff and patch dictionaries."""
import datetime
import decimal
import enum
import importlib
import pathlib
import uuid

from collections.abc import (Iterable, MutableMapping, MutableSequence,
                             MutableSet)
from copy import deepcopy

from .utils import EPSILON, PathLimit, are_different, dot_lookup
from .version import __version__

(ADD, REMOVE, CHANGE) = (
    'add', 'remove', 'change')

__all__ = ('diff', 'patch', 'swap', 'revert', 'dot_lookup', 'add_transform', 'allow_import', '__version__')

DICT_TYPES = (MutableMapping, )
LIST_TYPES = (MutableSequence, )
SET_TYPES = (MutableSet, )
ALL_TYPES = DICT_TYPES + LIST_TYPES + SET_TYPES

TRANSFORMS = [
    (datetime.datetime, {'from': lambda value: value.isoformat(), 'to': datetime.datetime.fromisoformat}),
    (datetime.date, {'from': lambda value: value.isoformat(), 'to': datetime.date.fromisoformat}),
    (datetime.time, {'from': lambda value: value.isoformat(), 'to': datetime.time.fromisoformat}),
    (datetime.timedelta, {
        'from': lambda value: value.total_seconds(),
        'to': lambda value: datetime.timedelta(seconds=value)
    }),
    (decimal.Decimal, {'from': str, 'to': decimal.Decimal}),
    (enum.Enum, {'from': lambda value: value.value, 'to': lambda value: value}),
    (pathlib.Path, {'from': str, 'to': pathlib.Path}),
    (uuid.UUID, {'from': str, 'to': uuid.UUID}),
]
VALUE_KEY = '_dictdiffer_value_key'
ALLOW_IMPORT = ['types', 'dataclasses']

try:
    import numpy
    HAS_NUMPY = True
    LIST_TYPES += (numpy.ndarray, )
except ImportError:  # pragma: no cover
    HAS_NUMPY = False


def diff(first, second, node=None, ignore=None, path_limit=None, expand=False,
         tolerance=EPSILON, absolute_tolerance=None, dot_notation=True):
    """Compare two dictionary/list/set objects, and returns a diff result.

    Return an iterator with differences between two objects. The diff items
    represent addition/deletion/change and the item value is a *deep copy*
    from the corresponding source or destination objects.

    >>> from dictdiffer import diff
    >>> result = diff({'a': 'b'}, {'a': 'c'})
    >>> list(result)
    [('change', 'a', ('b', 'c'))]

    The keys can be skipped from difference calculation when they are included
    in ``ignore`` argument of type :class:`collections.Container`.

    >>> list(diff({'a': 1, 'b': 2}, {'a': 3, 'b': 4}, ignore=set(['a'])))
    [('change', 'b', (2, 4))]
    >>> class IgnoreCase(set):
    ...     def __contains__(self, key):
    ...         return set.__contains__(self, str(key).lower())
    >>> list(diff({'a': 1, 'b': 2}, {'A': 3, 'b': 4}, ignore=IgnoreCase('a')))
    [('change', 'b', (2, 4))]

    The difference calculation can be limited to certain path:

    >>> list(diff({}, {'a': {'b': 'c'}}))
    [('add', '', [('a', {'b': 'c'})])]

    >>> from dictdiffer.utils import PathLimit
    >>> list(diff({}, {'a': {'b': 'c'}}, path_limit=PathLimit()))
    [('add', '', [('a', {})]), ('add', 'a', [('b', 'c')])]

    >>> from dictdiffer.utils import PathLimit
    >>> list(diff({}, {'a': {'b': 'c'}}, path_limit=PathLimit([('a',)])))
    [('add', '', [('a', {'b': 'c'})])]

    >>> from dictdiffer.utils import PathLimit
    >>> list(diff({}, {'a': {'b': 'c'}},
    ...           path_limit=PathLimit([('a', 'b')])))
    [('add', '', [('a', {})]), ('add', 'a', [('b', 'c')])]

    The patch can be expanded to small units e.g. when adding multiple values:

    >>> list(diff({'fruits': []}, {'fruits': ['apple', 'mango']}))
    [('add', 'fruits', [(0, 'apple'), (1, 'mango')])]

    >>> list(diff({'fruits': []}, {'fruits': ['apple', 'mango']}, expand=True))
    [('add', 'fruits', [(0, 'apple')]), ('add', 'fruits', [(1, 'mango')])]

    >>> list(diff({'a': {'x': 1}}, {'a': {'x': 2}}))
    [('change', 'a.x', (1, 2))]

    >>> list(diff({'a': {'x': 1}}, {'a': {'x': 2}},
    ... dot_notation=False))
    [('change', ['a', 'x'], (1, 2))]

    Diffing treats objects with ``__dict__`` as ``dict``s:

    >>> from types import SimpleNamespace as obj
    >>> list(diff(obj(a=1), obj(a=2)))
    [('change', ['__dict__', 'a'], (1, 2))]

    :param first: The original dictionary, ``list`` or ``set``.
    :param second: New dictionary, ``list`` or ``set``.
    :param node: Key for comparison that can be used in :func:`dot_lookup`.
    :param ignore: Set of keys that should not be checked.
    :param path_limit: List of path limit tuples or dictdiffer.utils.Pathlimit
                       object to limit the diff recursion depth.
    :param expand: Expand the patches.
    :param tolerance: Threshold to consider when comparing two float numbers.
    :param absolute_tolerance: Absolute threshold to consider when comparing
                               two float numbers.
    :param dot_notation: Boolean to toggle dot notation on and off.

    In general:

    - `diff` is a generator that yields zero or more tuples of the format
    `(op, path, values)`, where
    - `op` is one of the strings 'add', 'change' or 'remove'
    - `path` is by default a dot-separated string of keys from the root of the
    structure to the point of difference.
        - If parameter `dot_notation` is set to False, path is a list of
        separate key strings instead.
    - `values` are one or more tuples containing:
        - Key/value pairs for 'add' and 'remove'; keys for lists are indexes
        - Previous/new values for 'change', key being a part of the path in this case
        - Several value tuples sharing the same op and path are wrapped in a
        list, unless you specify `expand=True`, in which case they all
        get a separate (op, path, values) tuple.

    .. versionchanged:: 0.3
       Added *ignore* parameter.

    .. versionchanged:: 0.4
       Arguments ``first`` and ``second`` can now contain a ``set``.

    .. versionchanged:: 0.5
       Added *path_limit* parameter.
       Added *expand* paramter.
       Added *tolerance* parameter.

    .. versionchanged:: 0.7
       Diff items are deep copies from its corresponding objects.
       Argument *ignore* is always converted to a ``set``.

    .. versionchanged:: 0.8
        Added *dot_notation* parameter.
    """
    if path_limit is not None and not isinstance(path_limit, PathLimit):
        path_limit = PathLimit(path_limit)

    if isinstance(ignore, Iterable):
        def _process_ignore_value(value):
            if isinstance(value, int):
                return value,
            elif isinstance(value, list):
                return tuple(value)
            elif not dot_notation and isinstance(value, str):
                return value,
            return value

        ignore = type(ignore)(_process_ignore_value(value) for value in ignore)

    def dotted(node, default_type=list):
        """Return dotted notation."""
        if dot_notation and \
            all(map(lambda x: isinstance(x, str) and '.' not in x,
                node)):
            return '.'.join(node)
        else:
            return default_type(node)

    def _diff_recursive(_first, _second, _node=None):
        _node = _node or []

        if (
            not isinstance(_first, ALL_TYPES) and hasattr(_first, '__dict__') and
            not isinstance(_second, ALL_TYPES) and hasattr(_second, '__dict__')
        ):
            _first = _first.__dict__
            _second = _second.__dict__
            _node = _node + ['__dict__']

        dotted_node = dotted(_node)

        differ = False

        if isinstance(_first, DICT_TYPES) and isinstance(_second, DICT_TYPES):
            # dictionaries are not hashable, we can't use sets
            def check(key):
                """Test if key in current node should be ignored."""
                return ignore is None or (
                    dotted(_node + [key], default_type=tuple) not in ignore and
                    tuple(_node + [key]) not in ignore
                )

            intersection = [k for k in _first if k in _second and check(k)]
            addition = [k for k in _second if k not in _first and check(k)]
            deletion = [k for k in _first if k not in _second and check(k)]

            differ = True

        elif isinstance(_first, LIST_TYPES) and isinstance(_second,
                                                           LIST_TYPES):
            len_first = len(_first)
            len_second = len(_second)

            intersection = list(range(0, min(len_first, len_second)))
            addition = list(range(min(len_first, len_second), len_second))
            deletion = list(
                reversed(range(min(len_first, len_second), len_first)))

            differ = True

        elif isinstance(_first, SET_TYPES) and isinstance(_second, SET_TYPES):
            # Deep copy is not necessary for hashable items.
            addition = _second - _first
            if len(addition):
                yield ADD, dotted_node, [(0, addition)]
            deletion = _first - _second
            if len(deletion):
                yield REMOVE, dotted_node, [(0, deletion)]

            return  # stop here for sets

        if differ:
            # Compare if object is a dictionary or list.
            #
            # NOTE variables: intersection, addition, deletion contain only
            # hashable types, hence they do not need to be deepcopied.
            #
            # Call again the parent function as recursive if dictionary have
            # child objects.  Yields `add` and `remove` flags.
            for key in intersection:
                # if type is not changed,
                # calls again diff function to compare.
                # otherwise, the change will be handled as `change` flag.
                if path_limit and path_limit.path_is_limit(_node + [key]):
                    yield CHANGE, _node + [key], (
                        represent(_first[key]), represent(_second[key])
                    )
                else:
                    recurred = _diff_recursive(
                        _first[key], _second[key],
                        _node=_node + [key],
                    )

                    for diffed in recurred:
                        yield diffed

            if addition:
                if path_limit:
                    collect = []
                    collect_recurred = []
                    for key in addition:
                        if not isinstance(_second[key], ALL_TYPES):
                            collect.append((key, represent(_second[key])))
                        elif path_limit.path_is_limit(_node + [key]):
                            collect.append((key, represent(_second[key])))
                        else:
                            collect.append((key, _second[key].__class__()))
                            recurred = _diff_recursive(
                                _second[key].__class__(),
                                _second[key],
                                _node=_node + [key],
                            )

                            collect_recurred.append(recurred)

                    if expand:
                        for key, val in collect:
                            yield ADD, dotted_node, [(key, val)]
                    else:
                        yield ADD, dotted_node, collect

                    for recurred in collect_recurred:
                        for diffed in recurred:
                            yield diffed
                else:
                    if expand:
                        for key in addition:
                            yield ADD, dotted_node, [
                                (key, represent(_second[key]))]
                    else:
                        yield ADD, dotted_node, [
                            # for additions, return a list that consist with
                            # two-pair tuples.
                            (key, represent(_second[key])) for key in addition]

            if deletion:
                if expand:
                    for key in deletion:
                        yield REMOVE, dotted_node, [
                            (key, represent(_first[key]))]
                else:
                    yield REMOVE, dotted_node, [
                        # for deletions, return the list of removed keys
                        # and values.
                        (key, represent(_first[key])) for key in deletion]

        else:
            # Compare string and numerical types and yield `change` flag.
            if are_different(_first, _second, tolerance, absolute_tolerance):
                yield CHANGE, dotted_node, (represent(_first),
                                            represent(_second))

    return _diff_recursive(first, second, node)


def patch(diff_result, destination, in_place=False):
    """Patch the diff result to the destination dictionary.

    :param diff_result: Changes returned by ``diff``.
    :param destination: Structure to apply the changes to.
    :param in_place: By default, destination dictionary is deep copied
                     before applying the patch, and the copy is returned.
                     Setting ``in_place=True`` means that patch will apply
                     the changes directly to and return the destination
                     structure.
                     Note that patching is not atomic -
                     an exception in patching while ``in_place=True``
                     can leave the structure in a state where only a part of
                     the patch was applied.
    """
    if not in_place:
        destination = deepcopy(destination)

    def add(node, changes):
        for key, value in changes:
            dest = dot_lookup(destination, node)
            if isinstance(dest, LIST_TYPES):
                dest.insert(key, value)
            elif isinstance(dest, SET_TYPES):
                dest |= value
            else:
                dest[key] = reconstruct(value)

    def change(node, changes):
        dest = dot_lookup(destination, node, parent=True)
        if isinstance(node, str):
            last_node = node.split('.')[-1]
        else:
            last_node = node[-1]
        if isinstance(dest, LIST_TYPES):
            last_node = int(last_node)
        _, value = changes
        dest[last_node] = reconstruct(value)

    def remove(node, changes):
        for key, value in changes:
            dest = dot_lookup(destination, node)
            if isinstance(dest, SET_TYPES):
                dest -= value
            else:
                del dest[key]

    patchers = {
        REMOVE: remove,
        ADD: add,
        CHANGE: change
    }

    for action, node, changes in diff_result:
        patchers[action](node, changes)

    return destination


def swap(diff_result):
    """Swap the diff result.

    It uses following mapping:

    - remove -> add
    - add -> remove

    In addition, swap the changed values for `change` flag.

        >>> from dictdiffer import swap
        >>> swapped = swap([('add', 'a.b.c', [('a', 'b'), ('c', 'd')])])
        >>> next(swapped)
        ('remove', 'a.b.c', [('c', 'd'), ('a', 'b')])

        >>> swapped = swap([('change', 'a.b.c', ('a', 'b'))])
        >>> next(swapped)
        ('change', 'a.b.c', ('b', 'a'))

    """
    def add(node, changes):
        return REMOVE, node, list(reversed(changes))

    def remove(node, changes):
        return ADD, node, changes

    def change(node, changes):
        first, second = changes
        return CHANGE, node, (second, first)

    swappers = {
        REMOVE: remove,
        ADD: add,
        CHANGE: change
    }

    for action, node, change in diff_result:
        yield swappers[action](node, change)


def revert(diff_result, destination, in_place=False):
    """Call swap function to revert patched dictionary object.

    Usage example:

        >>> from dictdiffer import diff, revert
        >>> first = {'a': 'b'}
        >>> second = {'a': 'c'}
        >>> revert(diff(first, second), second)
        {'a': 'b'}

    :param diff_result: Changes returned by ``diff``.
    :param destination: Structure to apply the changes to.
    :param in_place: By default, destination dictionary is deep
                     copied before being reverted, and the copy
                     is returned. Setting ``in_place=True`` means
                     that revert will apply the changes directly to
                     and return the destination structure.
    """
    return patch(swap(diff_result), destination, in_place)


def represent(value):
    """
    Return object values such as decimal.Decimal or objects with a __dict__ member in a format that can be
    reconstructed in patching.

    >>> import decimal
    >>> represent(decimal.Decimal("1.23"))
    {'_dictdiffer_value_key': {'type': 'decimal.Decimal', 'value': '1.23'}}
    >>> import datetime
    >>> represent(datetime.date(2021, 7, 6))
    {'_dictdiffer_value_key': {'type': 'datetime.date', 'value': '2021-07-06'}}
    """
    transformed_value = False
    for cls, transform in TRANSFORMS:
        if issubclass(type(value), cls):
            transformed_value = transform['from'](value)
            represent_type = cls
            break
    else:
        represent_type = type(value)
        if represent_type.__module__ in ALLOW_IMPORT and hasattr(value, '__dict__'):
            transformed_value = value.__dict__

    if transformed_value:
        value = {'_dictdiffer_value_key': {
            'module': represent_type.__module__,
            'name': represent_type.__name__,
            'value': transformed_value
        }}

    return deepcopy(value)


def reconstruct(value):
    if type(value) is dict:
        value_spec = value.get('_dictdiffer_value_key')
        if value_spec:
            module_name = value_spec['module']
            class_name = value_spec['name']
            spec_value = value_spec['value']

            # Enums cannot be reconstructed, we just use the value
            if module_name == 'enum' and class_name == 'Enum':
                return spec_value

            # Try to match with defined basic types like dates, decimals
            for cls, transform in TRANSFORMS:
                if cls.__module__ == module_name and cls.__name__ == class_name:
                    value = transform['to'](spec_value)
                    break

            # Check if we can re-instantiate a class from an allowed module like types
            else:
                if module_name in ALLOW_IMPORT:
                    module = importlib.import_module(module_name)
                    cls = getattr(module, class_name)
                    value = cls(**spec_value)
                else:
                    raise ValueError(f'Could not reconstruct value {value}')
    return value


def add_transform(value_sample, represent, reconstruct):
    TRANSFORMS.append((type(value_sample), {'from': represent, 'to': reconstruct}))
    assert reconstruct(represent(value_sample)) == value_sample, (
        f'Could not reconstruct ({type(represent(value_sample)).__name__}) {represent(value_sample)} '
        f'to ({type(value_sample).__name__}) {value_sample}'
    )


def allow_import(*module_names):
    for module in module_names:
        importlib.import_module(module)
    ALLOW_IMPORT.extend(module_names)
